src/security_context.rs

```diff
@@ -71,7 +72,7 @@ impl SecurityContext {
                     Ok((SecurityContext(ctxt), Some(ContextBuffer(outbuf[0]))))
                 }
                 err => {
-                    Err(io::Error::from_raw_os_error(err as i32))
+                    Err((io::Error::from_raw_os_error(err as i32), SecurityContext(ctxt)))
```

Comment:
That's fair, but this is the only sensical place I could see to grab it because it seems that it needs to be initialized one way or another to hold meaningful failure information. The Handshake failure happens right here as far as I can tell. Maybe the SecurityContext isn't the actual object we want to be passing back in case of a failure?